src/plugins/usage_collection/server/collector/collector_set.ts

```diff
@@ -115,22 +112,40 @@ export class CollectorSet {
       );
     }
 
-    const secondInMs = 1000;
+    const timeoutMs = this.maximumWaitTimeForAllCollectorsInS * SECOND_IN_MS;
     const collectorsWithStatus: CollectorWithStatus[] = await Promise.all(
       [...collectors.values()].map(async (collector) => {
+        const wrappedPromise = performance.timerify(
+          Object.defineProperty(
+            async (): Promise<boolean> => {
+              try {
+                return await collector.isReady();
+              } catch (err) {
+                this.logger.debug(`Collector ${collector.type} failed to get ready. ${err}`);
+                return false;
+              }
+            },
+            'name',
+            { value: `is_ready_${collector.type}` }
+          )
+        );
+
         const isReadyWithTimeout = await withTimeout<boolean>({
-          promise: (async (): Promise<boolean> => {
-            try {
-              return await collector.isReady();
-            } catch (err) {
-              this.logger.debug(`Collector ${collector.type} failed to get ready. ${err}`);
-              return false;
-            }
-          })(),
-          timeoutMs: this.maximumWaitTimeForAllCollectorsInS * secondInMs,
+          promise: wrappedPromise(),
+          timeoutMs,
         });
 
-        return { isReadyWithTimeout, collector };
+        if (isReadyWithTimeout.timedout) {
+          return { isReadyWithTimeout, collector };
+        }
+
+        return {
+          isReadyWithTimeout: {
+            value: isReadyWithTimeout.value,
+            timedout: isReadyWithTimeout.timedout,
+          },
+          collector,
+        };
```

Comment:
I believe i did that some time ago because it was the easiest way to inform the TS compiler that there are 2 different types here (timedout: true has no value while timedout false has a value). I will check if it will bs smart about it after the proposed change in the new TS version.